### PR TITLE
Agent to inject rules to disable masquerade in Cilium

### PR DESCRIPTION
**Pull request by alex-hunt-materialize**
Adds a new binary for an agent to inject rules to disable masquerade for
pods with EIPs when using Cilium as the CNI.

Also creates a shared library crate for code shared between the two binaries.

### CODE REVIEW

File: cilium_eip_no_masquerade_agent/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+                && rule
+                    .nlas
+                    .contains(&Nla::Destination(vpc_cidr.network().octets().to_vec()))
+            {
+                Some(rule.header.table)
+            } else {
+                None
+            }
+        })
+        .ok_or(Error::CiliumRuleNotFound)?;
```
Comment (benesch) [inline cilium_eip_no_masquerade_agent/src/main.rs:89]: This causes it to get rescheduled later, right?

File: cilium_eip_no_masquerade_agent/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+use rand::{thread_rng, Rng};
+use rtnetlink::packet::rule::Nla;
+use rtnetlink::packet::RuleMessage;
+use rtnetlink::{new_connection, Handle, IpVersion, RuleHandle};
+use tracing::{debug, event, info, instrument, Level};
+
+use eip_operator_shared::{run_with_tracing, Error};
+
+const MANAGE_EIP_LABEL: &str = "eip.materialize.cloud/manage";
+const POD_FINALIZER_NAME: &str = "eip.materialize.cloud/cilium_no_masquerade_rule";
```
Comment (benesch) [inline cilium_eip_no_masquerade_agent/src/main.rs:21]:
```suggestion
const POD_FINALIZER_NAME: &str = "eip.materialize.cloud/cilium-no-masquerade-rule";
```

File: cilium_eip_no_masquerade_agent/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+use kube_runtime::finalizer::{finalizer, Event};
+use rand::{thread_rng, Rng};
+use rtnetlink::packet::rule::Nla;
+use rtnetlink::packet::RuleMessage;
+use rtnetlink::{new_connection, Handle, IpVersion, RuleHandle};
+use tracing::{debug, event, info, instrument, Level};
+
+use eip_operator_shared::{run_with_tracing, Error};
+
+const MANAGE_EIP_LABEL: &str = "eip.materialize.cloud/manage";
```
Comment (benesch) [inline cilium_eip_no_masquerade_agent/src/main.rs:20]: This should move to the shared crate, no?

File: cilium_eip_no_masquerade_agent/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    rules: impl TryStream<Ok = RuleMessage, Error = rtnetlink::Error>,
+    pod_ip: Ipv4Addr,
+) -> Result<Vec<RuleMessage>, rtnetlink::Error> {
+    rules
+        .try_filter(|rule| {
+            future::ready(
+                rule.header.src_len == 32
+                    && rule.nlas.contains(&Nla::Source(pod_ip.octets().to_vec())),
+            )
+        })
```
Comment (benesch) [inline cilium_eip_no_masquerade_agent/src/main.rs:49]:
```suggestion
        .try_filter(|rule| async {
                rule.header.src_len == 32
                    && rule.nlas.contains(&Nla::Source(pod_ip.octets().to_vec())),
            }
        })
```

**Review by benesch — APPROVED** [review]

File: cilium_eip_no_masquerade_agent/src/main.rs (hunk 1 above)
Comment (alex-hunt-materialize) [reply cilium_eip_no_masquerade_agent/src/main.rs:89]: Yes, it gets requeued for a random time between 4-8 seconds later.

File: cilium_eip_no_masquerade_agent/src/main.rs (hunk 4 above)
Comment (alex-hunt-materialize) [reply cilium_eip_no_masquerade_agent/src/main.rs:49]:
```
error: lifetime may not live long enough
  --> cilium_eip_no_masquerade_agent/src/main.rs:43:28
   |
43 |           .try_filter(|rule| async {
   |  ______________________-----_^
   | |                      |   |
   | |                      |   return type of closure `impl futures::Future<Output = bool>` contains a lifetime `'2`
   | |                      has type `&'1 RuleMessage`
44 | |             rule.header.src_len == 32 && rule.nlas.contains(&Nla::Source(pod_ip.octets().to_vec()))
45 | |         })
   | |_________^ returning this value requires that `'1` must outlive `'2`
```

Comment (benesch) [reply cilium_eip_no_masquerade_agent/src/main.rs:49]: Ugh. Forget I said anything.

**Pull request merged**